torch/csrc/autograd/engine.cpp

```diff
@@ -586,6 +586,7 @@ void GraphTask::exec_post_processing() {
     // added in between callback calls, so iterators may become invalidated.
     // NOLINTNEXTLINE(modernize-loop-convert)
     for (size_t i = 0; i < final_callbacks_.size(); ++i) {
+      at::ThreadLocalStateGuard guard(this->thread_locals_);
```

Comment:
can you move this outside of the for loop? There is no reason to set the same thing again and again in the loop.

Reply:
A new code block has been added to wrap the final callback looping.